File: languages/java/src/main/java/com/bloock/sdk/client/EncryptionClient.java
package com.bloock.sdk.client;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.BloockConfig.ConfigData;
import com.bloock.sdk.bridge.proto.BloockEncryption;
import com.bloock.sdk.bridge.proto.BloockEncryption.EncryptionAlgRequest;
import com.bloock.sdk.bridge.proto.BloockEncryption.EncryptionAlgResponse;
import com.bloock.sdk.bridge.proto.BloockKeys;
import com.bloock.sdk.bridge.proto.BloockShared.Error;
import com.bloock.sdk.config.Config;
import com.bloock.sdk.entity.encryption.Encrypter;
import com.bloock.sdk.entity.encryption.EncryptionAlg;
import com.bloock.sdk.entity.key.KeyType;
import com.bloock.sdk.entity.key.RsaKeyPair;
import com.bloock.sdk.entity.record.Record;

/**
 * Represents a client for interacting with the
 * <a href="https://dashboard.bloock.com/login">Bloock Encryption service</a>.
 */
public class EncryptionClient {
  private final Bridge bridge;
  private final ConfigData configData;

  /**
   * Creates a new instance of the EncryptionClient with default configuration.
   */
  public EncryptionClient() {
    this.bridge = new Bridge();
    this.configData = Config.newConfigDataDefault();
  }

  /**
   * Creates a new instance of the EncryptionClient with the provided
   * configuration.
   * 
   * @param configData
   */
  public EncryptionClient(ConfigData configData) {
    this.bridge = new Bridge();
    this.configData = Config.newConfigData(configData);
  }

  /**
   * GenerateRsaKeyPair generates an RSA key pair for encryption.
   * 
   * @deprecated Will be deleted in future versions. Use KeyClient.newLocalKey
   *             function instead.
   */
  @Deprecated
  public RsaKeyPair generateRsaKeyPair() throws Exception {
    BloockKeys.GenerateLocalKeyRequest request = BloockKeys.GenerateLocalKeyRequest.newBuilder()
        .setConfigData(this.configData)
        .setKeyType(KeyType.Rsa2048.toProto())
        .build();

    BloockKeys.GenerateLocalKeyResponse response = bridge.getKey().generateLocalKey(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return RsaKeyPair.fromProto(response);
  }

  /**
   * Encrypts a Bloock record using the specified encrypter.
   * 
   * @param record
   * @param encrypter
   * @return
   * @throws Exception
   */
  public Record encrypt(Record record, Encrypter encrypter) throws Exception {
    BloockEncryption.EncryptRequest request = BloockEncryption.EncryptRequest.newBuilder()
        .setConfigData(this.configData)
        .setRecord(record.toProto())
        .setEncrypter(encrypter.toProto())
        .build();

    BloockEncryption.EncryptResponse response = bridge.getEncryption().encrypt(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Record.fromProto(response.getRecord(), this.configData);
  }

  /**
   * Decrypts a Bloock record using the specified decrypter.
   * 
   * @param record
   * @param decrypter
   * @return
   * @throws Exception
   */
  public Record decrypt(Record record, Encrypter decrypter) throws Exception {
    BloockEncryption.DecryptRequest request = BloockEncryption.DecryptRequest.newBuilder()
        .setConfigData(this.configData)
        .setRecord(record.toProto())
        .setDecrypter(decrypter.toProto())
        .build();

    BloockEncryption.DecryptResponse response = bridge.getEncryption().decrypt(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Record.fromProto(response.getRecord(), this.configData);
  }

  /**
   * Gets the encryption algorithm used for a Bloock record.
   * 
   * @param record
   * @return
   * @throws Exception
   */
  public EncryptionAlg getEncryptionAlg(Record record) throws Exception {
    Bridge bridge = new Bridge();

    EncryptionAlgRequest req = EncryptionAlgRequest.newBuilder()
        .setConfigData(this.configData)
        .setRecord(record.toProto())
        .build();
    EncryptionAlgResponse res = bridge.getEncryption().getEncryptionAlg(req);

    if (res.getError() != Error.getDefaultInstance()) {
      throw new Exception(res.getError().getMessage());
    }

    return EncryptionAlg.fromProto(res.getAlg());
  }
}

File: languages/php/Bloock/Client/EncryptionClient.php
<?php

namespace Bloock\Client;

use Bloock\Bridge\Bridge;
use Bloock\Config\Config;
use Bloock\ConfigData;
use Bloock\DecryptRequest;
use Bloock\EncryptionAlgRequest;
use Bloock\EncryptRequest;
use Bloock\Entity\Encryption\Encrypter;
use Bloock\Entity\Encryption\EncryptionAlg;
use Bloock\Entity\Key\RsaKeyPair;
use Bloock\Entity\Record\Record;
use Bloock\GenerateLocalKeyRequest;
use Bloock\KeyType;
use Exception;

/**
 * Represents a client for interacting with the [Bloock Encryption service](https://dashboard.bloock.com/login).
 */
class EncryptionClient
{
    private $bridge;
    private $config;

    /**
     * Creates a new instance of the EncryptionClient with the provided configuration.
     * @param ConfigData|null $config
     */
    public function __construct(?ConfigData $config = null)
    {
        $this->bridge = new Bridge();
        if ($config != null) {
            $this->config = Config::newConfigData($config);
        } else {
            $this->config = Config::newConfigDataDefault();
        }
    }

    /**
     * GenerateRsaKeyPair generates an RSA key pair for encryption.
     * @deprecated Will be deleted in future versions. Use KeyClient.newLocalKey function instead.
     */
    public function generateRsaKeyPair(): RsaKeyPair
    {
        $req = new GenerateLocalKeyRequest();
        $req->setConfigData($this->config);
        $req->setKeyType(KeyType::Rsa2048);

        $res = $this->bridge->key->GenerateLocalKey($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return RsaKeyPair::fromProto($res);
    }

    /**
     * Encrypts a Bloock record using the specified encrypter.
     * @param Record $record
     * @param Encrypter $encrypter
     * @return Record
     * @throws Exception
     */
    public function encrypt(Record $record, Encrypter $encrypter): Record
    {
        $req = new EncryptRequest();
        $req->setConfigData($this->config)->setRecord($record->toProto())->setEncrypter($encrypter->toProto());

        $res = $this->bridge->encryption->Encrypt($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Record::fromProto($res->getRecord(), $this->config);
    }

    /**
     * Decrypts a Bloock record using the specified decrypter.
     * @param Record $record
     * @param Encrypter $decrypter
     * @return Record
     * @throws Exception
     */
    public function decrypt(Record $record, Encrypter $decrypter): Record
    {
        $req = new DecryptRequest();
        $req->setConfigData($this->config)->setRecord($record->toProto())->setDecrypter($decrypter->toProto());

        $res = $this->bridge->encryption->Decrypt($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Record::fromProto($res->getRecord(), $this->config);
    }

    /**
     * Gets the encryption algorithm used for a Bloock record.
     * @param Record $record
     * @return string
     * @throws Exception
     */
    public function getEncryptionAlg(Record $record): string
    {
        $req = new EncryptionAlgRequest();
        $req->setConfigData($this->config)->setRecord($record->toProto());

        $res = $this->bridge->encryption->GetEncryptionAlg($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return EncryptionAlg::fromProto($res->getAlg());
    }
}
